### PR TITLE
feat(tile-converter): support 3tz

### PR DESCRIPTION
**Pull request by belom88**

### CODE REVIEW

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  FileProvider,
+  ZipFileSystem,
+  cdSignature as cdHeaderSignature,
+  searchFromTheEnd,
+  parseZipCDFileHeader,
+  HashElement,
+  parseHashFile,
+  parseZipLocalFileHeader
+} from '@loaders.gl/zip';
+import {TZ3Archive} from '../../tz3/tz3-archive';
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:11]: Nit: I personally think we should just drop the tz3 name completely. Given that it is a scrambled version of `3tz` it is just too obscure.
```suggestion
import {Tiles3DArchive} from '../../tiles-3d-archive/tiles-3d-archive';
```

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    const fileProvider = await this.fileProvider;
+    if (!fileProvider) {
+      throw new Error('No data detected in the zip archive');
+    }
+
+    const hashCDOffset = await searchFromTheEnd(fileProvider, cdHeaderSignature);
+
+    const cdFileHeader = await parseZipCDFileHeader(hashCDOffset, fileProvider);
+
+    if (cdFileHeader?.fileName === '@3dtilesIndex1@') {
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:69]: Some comments here perhaps?

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+   * @returns - Response with file data
+   */
+  async fetch(filename: string): Promise<Response> {
+    const fileProvider = await this.fileProvider;
+    if (!fileProvider) {
+      throw new Error('No data detected in the zip archive');
+    }
+    await this.parseHashFile();
+    if (this.hashData) {
+      const archive = new TZ3Archive(fileProvider, this.hashData);
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:40]: This looks odd. We are creating one archive on every fetch? Are there multiple archives in here?

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  /**
+   * Constructor
+   * @param file - instance of FileProvider or file path string
+   */
+  constructor(file: FileProvider | string) {
+    super(file);
+  }
+
+  /**
+   * Implementation of fetch against this file system
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:29]: Maybe instead describe the unique things that fetch needs to do on this file system

File: modules/tile-converter/src/i3s-converter/helpers/load-3d-tiles.ts [hunk 5]
```diff
@@ -62,7 +67,53 @@ export const loadTile3DContent = async (
...
 };
+
+/**
+ * Load a resrource with load options and .3tz format support
+ * @param url - resource URL
+ * @param loader - loader to parse data (Tiles3DLoader / CesiumIonLoader)
+ * @param loadOptions - 3d-tiles loader options
+ * @returns 3d-tiles resource
+ */
+export async function loadWithOptions(
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/helpers/load-3d-tiles.ts:86]: We normally just call this `load`, all load call have options. Maybe update name to indicate what we are loading instead, or what we are loading from?

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 6]
```diff
@@ -67,7 +67,12 @@ import {
   getFieldAttributeType
 } from './helpers/feature-attributes';
 import {NodeIndexDocument} from './helpers/node-index-document';
-import {loadNestedTileset, loadTile3DContent} from './helpers/load-3d-tiles';
+import {
+  isTilesetType,
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/i3s-converter.ts:71]: I find the name confusing. `isTileset` makes some sense, but `isTilesetType` is a bit much.

How about
```typescript
getTilesetType(tile): 'tileset' | 'tile'

if (getTilesetType(tile) === 'tileset') 
```
`

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 7]
```diff
@@ -562,8 +568,9 @@ export default class I3SConverter {
     sourceTile: Tiles3DTileJSONPostprocessed,
     traversalProps: TraversalConversionProps
   ): Promise<TraversalConversionProps> {
-    if (sourceTile.type === 'json' || sourceTile.type === 'empty') {
-      if (sourceTile.type === 'json') {
+    const isTileset = isTilesetType(sourceTile.type);
+    if (isTileset || sourceTile.type === 'empty') {
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/i3s-converter.ts:572]: Nit: This logic is very nested...

**Review by ibgreen — APPROVED** [review]

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts (hunk 3 above)
Comment (belom88) [reply modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:40]: It is the same. We create it to use `getFile` method

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts (hunk 1 above)
Comment (belom88) [reply modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:11]: Will be done in another PR

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts (hunk 2 above)
Comment (belom88) [reply modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts:77]: Done

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts (hunk 4 above)
Comment (belom88) [reply modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts]: Done

File: modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  parseZipCDFileHeader,
+  HashElement,
+  parseHashFile,
+  parseZipLocalFileHeader
+} from '@loaders.gl/zip';
+import {TZ3Archive} from '../../tz3/tz3-archive';
+
+/**
+ * FileSystem adapter for a 3tz (3D tiles archive format) file
+ * Holds FileProvider object that provides random access to archived files
```
[the remark this answers is not in the document]
Comment (belom88) [reply modules/3d-tiles/src/lib/filesystems/tiles-3d-archive-file-system.ts]: Done

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts [hunk 9]
```diff
@@ -89,7 +89,40 @@ function resolveUri(uri: string = '', basePath: string): string {
...
+/**
+ * Concatenate relative paths properly
+ * @param basePath base path of the file
+ * @param uri relative path of the file
+ * @returns concatenated string
+ * @example if `basePath`==='/a/b/c' and `uri`==='../d', the result will be 'a/b/d'
+ * @example if `basePath`==='/a/b/c' and `uri`==='./d', the result will be 'a/b/c/d'
+ * @example if `basePath`==='/a/b/c/' and `uri`==='../../d', the result will be 'a/d'
+ */
+function concatenateRelativePaths(basePath: string, uri: string): string {
```
[the remark this answers is not in the document]
Comment (belom88) [reply modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts]: Done

File: modules/tile-converter/src/i3s-converter/helpers/load-3d-tiles.ts (hunk 5 above)
Comment (belom88) [reply modules/tile-converter/src/i3s-converter/helpers/load-3d-tiles.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts (hunk 6 above)
Comment (belom88) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

**Pull request merged**